src/scalib_ext/scalib/src/sasca/factor_graph.rs

```diff
@@ -343,18 +345,83 @@ impl FactorGraph {
-                    // intentionally left empty).
+                FactorKind::GenFactor { id, operands } => {
+                    let ops: Vec<&PublicValue> = operands
+                        .iter()
+                        .map(|op| match op {
+                            GenFactorOperand::Var(idx, ..) => &var_assignments[*idx],
+                            GenFactorOperand::Pub(idx) => &public_values[*idx],
+                        })
+                        .collect();
+                    let nmulti_ops = ops.iter().find_map(|op| {
+                        if let PublicValue::Multi(x) = op {
+                            Some(x.len())
+                        } else {
+                            None
+                        }
+                    });
+                    let nmulti_factor = if let super::GenFactor::Multi(gfv) = &gen_factors[*id] {
+                        Some(gfv.len())
+                    } else {
+                        None
+                    };
+                    let nmulti = match (nmulti_ops, nmulti_factor) {
+                        (Some(nm_ops), Some(nm_factors)) if nm_ops == nm_factors => nm_factors,
+                        (Some(nm_ops), Some(nm_factors)) => {
+                            return Err(FGError::InvalidGenericFactorAssignment(format!("Mismatch between multi declaration of GenFactor operands {} and GenFactor {}", nm_ops, nm_factors)))
+                        }
+                        (Some(nmulti), None) | (None, Some(nmulti)) => nmulti,
+                        (None, None) => 1,
+                    };
+                    for i in 0..nmulti {
+                        let gf = match &gen_factors[*id] {
+                            super::GenFactor::Single(gfs) => gfs,
+                            super::GenFactor::Multi(gfv) => &gfv[i],
+                        };
+                        let indices: Vec<ClassVal> = ops
+                            .iter()
+                            .map(|pv| match *pv {
+                                PublicValue::Single(x) => *x,
+                                PublicValue::Multi(xvec) => xvec[i],
+                            })
+                            .collect();
+                        match gf {
+                            super::GenFactorInner::Dense(dense_factor) => {
+                                if !(dense_factor[indices
+                                    .iter()
+                                    .map(|x| *x as usize)
+                                    .collect::<Vec<usize>>()
+                                    .as_slice()]
+                                    > 0.0)
+                                {
+                                    return Err(FGError::InvalidGenericFactorAssignment(format!(
+                                        "Invalid assignment to {}: {:?}",
+                                        factor_name.clone(),
+                                        indices.clone()
```

Comment:
The error message is not the most explicit. Would be good to say "factor {}", and what are the indices.

Also, ideally we'd avoid using the Debug representation for user-facing errors (not a big deal, if the printed string ends up being what we want).